platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java

```diff
@@ -1011,13 +1047,91 @@ public Stream<VirtualHashRecord> dirtyHashesForFlush(final long lastLeafPath) {
+     * size is lower than flush threshold, the copy is not flushed to disk, but rather
+     * its {@link #garbageCollect(long, long)} is called.
      *
      * @return
-     *        Estimated number of dirty internal nodes
+     *        Estimated number of dirty hashes
      */
-    public long estimatedInternalsCount(final long firstLeafPath) {
-        return (dirtyHashes == null) ? 0 : dirtyHashes.size();
+    public long estimatedHashesCount() {
+        return (dirtyHashes == null) ? 0 : (dirtyHashes.size() - filteredHashesCount.get());
+    }
+
+    /**
+     * This cache copy is selected to flush to disk, either because it is explicitly
+     * marked as such, or when cache copy estimated size exceeds flush threshold. Before
+     * flush, this method is used to mark redundant mutations as "filtered", so they aren't
+     * included to the streams for the data source. It may happen that after filtering the
+     * cache copy is no longer needed to flush. In this case, all filtered mutations are
+     * removed using {@link #garbageCollect(long, long)} method.
+     *
+     * <p>This method can only be called on sealed caches.
+     *
+     * @throws MutabilityException if called on a non-sealed cache instance.
+     */
+    public void prepareForFlush() {
+        if (!hashesAreImmutable.get() || !leafIndexesAreImmutable.get()) {
+            throw new MutabilityException("Cannot prepare for flushing for a non-sealed cache");
+        }
+        if (preparedForFlush.getAndSet(true)) {
+            throw new IllegalStateException("This cache has been already prepared for flush");
+        }
+        // Mark obsolete mutations to filter later and update "filtered" counters. These
+        // counters will affect the estimated size
+        final long version = getFastCopyVersion();
+        final long lastReleasedVersion = lastReleased.get();
+        filterMutations(
+                dirtyHashes, pathToDirtyHashIndex, version, lastReleasedVersion, filteredHashesCount, virtualMapConfig);
+        filterMutations(
+                dirtyLeafPaths,
+                pathToDirtyLeafIndex,
+                version,
+                lastReleasedVersion,
+                filteredLeafPathsCount,
+                virtualMapConfig);
+        filterMutations(
+                dirtyLeaves, keyToDirtyLeafIndex, version, lastReleasedVersion, filteredLeavesCount, virtualMapConfig);
+    }
+
+    /**
+     * If a cache copy's estimated size exceeded flush threshold, but after mutations are
+     * filtered the size drops below the threshold, the copy is not flushed to disk, but
+     * just removes all filtered mutations using this method.
+     *
+     * <p>This method can only be called on sealed caches.
+     *
+     * @throws MutabilityException if called on a non-sealed cache instance.
+     */
+    public void garbageCollect(final long firstLeafPath, final long lastLeafPath) {
```

Comment:
`firstLeafPath`, `lastLeafPath` are not used in the method.

Reply:
Removed